src/expertsystem/amplitude/sympy.py

```diff
@@ -0,0 +1,428 @@
+        helicity_list.append((particle, spin_projection))
+
+    # in order to ensure correct naming of amplitude coefficients the list has
+    # to be sorted by name. The same coefficient names have to be created for
+    # two graphs that only differ from a kinematic standpoint
+    # (swapped external edges)
+    return sorted(helicity_list, key=lambda entry: entry[0].name)
+
+
+def _generate_particles_string(
+    helicity_list: List[ParticleWithSpin],
+    use_helicity: bool = True,
+    make_parity_partner: bool = False,
+) -> str:
+    output_string = ""
+    for particle, spin_projection in helicity_list:
+        if particle.latex is not None:
+            output_string += particle.latex
+        else:
+            output_string += particle.name
+        if use_helicity:
+            if make_parity_partner:
+                helicity = -1 * spin_projection
+            else:
+                helicity = spin_projection
+            if helicity > 0:
+                helicity_str = f"+{helicity}"
+            else:
+                helicity_str = str(helicity)
+            output_string += f"_{{{helicity_str}}}"
+        output_string += " "
+    return output_string[:-1]
+
+
+class SympyHelicityAmplitudeGenerator:
+    """Amplitude model generator for the helicity formalism."""
+
+    def __init__(self, reaction_result: Result) -> None:
+        self.name_generator = _SympyHelicityAmplitudeNameGenerator()
+        self.__graphs = reaction_result.solutions
+        if len(self.__graphs) < 1:
+            raise ValueError(
+                f"At least one {StateTransitionGraph.__name__} required to"
+                " genenerate an amplitude model!"
+            )
+
+        get_initial_state = reaction_result.get_initial_state()
+        if len(get_initial_state) != 1:
+            raise ValueError(
+                "Helicity amplitude model requires exactly one initial state"
+            )
+        particles = generate_particle_collection(self.__graphs)
+        kinematics = generate_kinematics(reaction_result)
+        self.__model = SympyModel(
+            particles=particles,
+            kinematics=kinematics,
+        )
+
+    def generate(self) -> sy.Expr:
+        self.__model.intensity_model = self.__generate_intensities()
+        return self.__model
+
+    def __generate_intensities(self) -> sy.Expr:
+        graph_groups = group_graphs_same_initial_and_final(self.__graphs)
+        logging.debug("There are %d graph groups", len(graph_groups))
+
+        self.__create_parameter_couplings(graph_groups)
+        coherent_intensities = [
+            self.__generate_coherent_intensity(graph_group)
+            for graph_group in graph_groups
+        ]
+        if len(coherent_intensities) == 0:
+            raise ValueError("List of coherent intensities cannot be empty")
+        if len(coherent_intensities) == 1:
+            return coherent_intensities[0]
+        return sum(coherent_intensities)
```

Comment:
Shouldn't this be a sy.sum call? There is another one below.

Reply:
Afaik `sy.sum`/`sy.Sum` is for indexed symbols. Here we just have a sum of a list of `sy.Expr`. See also the use of `reduce` for sequential amplitudes elsewhere.